canopen/source/co_csdo.c

```diff
@@ -450,11 +495,392 @@ void COCSdoTransferFinalize(CO_CSDO *csdo) {
         csdo->Tfer.Tmt   = 0;
         csdo->Tfer.Call  = 0;
         csdo->Tfer.Tmr   = -1;
+        csdo->Tfer.Buf_Idx = 0;
+        csdo->Tfer.TBit = 0;
 
         /* Release SDO client for next request */
         csdo->Frm   = 0;
         csdo->State = CO_CSDO_STATE_IDLE;
     }
 }
 
+CO_ERR COCSdoInitUploadSegmented(CO_CSDO *csdo)
+{
+    CO_ERR result = CO_ERR_SDO_SILENT;
+    uint32_t obj_size;
+    uint32_t ticks;
+    uint16_t Idx;
+    uint8_t Sub;
+    CO_IF_FRM frm;
+    
+
+    obj_size = CO_GET_LONG(csdo->Frm, 4);
+    Idx = CO_GET_WORD(csdo->Frm, 1);
+    Sub = CO_GET_BYTE(csdo->Frm, 3);
+
+    /* verify size,Idx, Sub */
+    if ((obj_size == csdo->Tfer.Size) &&
+        (Idx == csdo->Tfer.Idx) &&
+        (Sub == csdo->Tfer.Sub))
+    {
+        result = CO_ERR_NONE;
+
+        /* setup CAN request */
+        CO_SET_ID  (&frm, csdo->TxId       );
+        CO_SET_DLC (&frm, 8                );
+        CO_SET_BYTE(&frm, 0x60         , 0);
+        CO_SET_WORD(&frm, 0, 1);
+        CO_SET_BYTE(&frm, 0, 3);
+        CO_SET_LONG(&frm, 0, 4);
+
+        csdo->Tfer.TBit = 0;
+
+        /* refresh timer */
+        COTmrDelete(&(csdo->Node->Tmr), csdo->Tfer.Tmr);
+        ticks = COTmrGetTicks(&(csdo->Node->Tmr), csdo->Tfer.Tmt, CO_TMR_UNIT_1MS);
+        csdo->Tfer.Tmr = COTmrCreate(&(csdo->Node->Tmr), ticks, 0, &COCSdoTimeout, csdo);
+
+        (void)COIfCanSend(&csdo->Node->If, &frm);
+    }
+    else
+    {
+        COCSdoAbort(csdo, CO_SDO_ERR_MEM);
```

Comment:
To get the correct abort code, we should use `CO_SDO_ERR_LEN`.

and we need to add this error code in co_sdo.h:

```c
#define CO_SDO_ERR_LEN          0x06070010    /*!< Length of parameter does not match     */
```